lib/backend/pgbk/background.go

```diff
@@ -202,42 +202,77 @@ func (b *Backend) pollChangeFeed(ctx context.Context, conn *pgx.Conn, slotName s
 	// TODO(espadolini): it might be better to do the JSON deserialization
 	// (potentially with additional checks for the schema) on the auth side
 	rows, _ := conn.Query(ctx,
-		`WITH jdata AS (
+		`WITH d AS (
   SELECT
     data::jsonb AS data
   FROM pg_logical_slot_get_changes($1, NULL, $2,
     'format-version', '2', 'add-tables', 'public.kv', 'include-transaction', 'false')
 )
 SELECT
-  data->>'action',
-  decode(COALESCE(data->'columns'->0->>'value', data->'identity'->0->>'value'), 'hex'),
-  decode(data->'columns'->1->>'value', 'hex'),
-  (data->'columns'->2->>'value')::timestamptz,
-  (data->'columns'->3->>'value')::uuid
-FROM jdata`, slotName, b.cfg.ChangeFeedBatchSize)
+  d.data->>'action' AS action,
+  decode(jsonb_path_query_first(d.data, '$.columns[*]?(@.name == "key")')->>'value', 'hex') AS key,
+  NULLIF(
+    decode(jsonb_path_query_first(d.data, '$.identity[*]?(@.name == "key")')->>'value', 'hex'),
+    decode(jsonb_path_query_first(d.data, '$.columns[*]?(@.name == "key")')->>'value', 'hex')
+  ) AS old_key,
+  decode(COALESCE(
+    jsonb_path_query_first(d.data, '$.columns[*]?(@.name == "value")'),
+    jsonb_path_query_first(d.data, '$.identity[*]?(@.name == "value")')
+  )->>'value', 'hex') AS value,
+  (COALESCE(
+    jsonb_path_query_first(d.data, '$.columns[*]?(@.name == "expires")'),
+    jsonb_path_query_first(d.data, '$.identity[*]?(@.name == "expires")')
+  )->>'value')::timestamptz AS expires,
+  (COALESCE(
+    jsonb_path_query_first(d.data, '$.columns[*]?(@.name == "revision")'),
+    jsonb_path_query_first(d.data, '$.identity[*]?(@.name == "revision")')
+  )->>'value')::uuid AS revision
```

Comment:
The `COALESCE` is giving me a pause. If the new value is missing, we use the old one as part of the update. Can we end up mixing the values from old and new records? Should we, perhaps, only ever use the new record here?

Reply:
I'll add comments to describe this, but the whole point is that in an UPDATE we might have some missing columns from the `columns` field, which happens if the column was a TOASTed column that was not updated - as opposed to the object being there but with a `"value":null` in the json sense, which means that the SQL value is known to be a NULL.

In INSERT there's only `columns`, in DELETE there's only `identity`, so coalesce and then accessing the `value` field gives us the correct result in all three cases.

Reply:
Thanks, this makes sense. I'd appreciate the comment here.